ntp-daemon/src/peer.rs

```diff
@@ -1,177 +1,277 @@
-    let our_id = ReferenceId::from_ip(socket.as_ref().local_addr().unwrap().ip());
-    let peer_id = ReferenceId::from_ip(socket.as_ref().peer_addr().unwrap().ip());
-    let (tx, rx) = watch::channel::<MsgForSystem>(MsgForSystem::NoMeasurement);
+    socket: UdpSocket,
+    channels: PeerChannels,
 
-    // channel to notify that a reset has been completed by this peer
-    let (notify_reset_send, notify_reset_receive) = watch::channel::<u64>(0);
+    peer: Peer,
 
-    tokio::spawn(async move {
-        let local_clock_time = NtpInstant::now();
-        let mut peer = Peer::new(our_id, peer_id, local_clock_time);
+    // we don't store the real origin timestamp in the packet, because that would leak our
+    // system time to the network (and could make attacks easier). So instead there is some
+    // garbage data in the origin_timestamp field, and we need to track and pass along the
+    // actual origin timestamp ourselves.
+    /// Timestamp of the last packet that we sent
+    last_send_timestamp: Option<NtpTimestamp>,
 
-        let poll_wait = tokio::time::sleep(std::time::Duration::default());
-        tokio::pin!(poll_wait);
+    /// Instant last poll message was sent (used for timing the wait)
+    last_poll_sent: Instant,
 
-        // we don't store the real origin timestamp in the packet, because that would leak our
-        // system time to the network (and could make attacks easier). So instead there is some
-        // garbage data in the origin_timestamp field, and we need to track and pass along the
-        // actual origin timestamp ourselves.
-        let mut last_send_timestamp = None;
+    /// Number of resets that this peer has performed
+    reset_epoch: ResetEpoch,
+}
 
-        // Instant last poll message was sent (used for timing the wait)
-        let mut last_poll_sent = Instant::now();
+impl<C> PeerTask<C>
+where
+    C: 'static + NtpClock + Send,
+{
+    /// Set the next deadline for the poll interval based on current state
+    fn update_poll_wait(&self, poll_wait: &mut Pin<&mut Sleep>, system_snapshot: SystemSnapshot) {
+        let poll_interval = self
+            .peer
+            .current_poll_interval(system_snapshot)
+            .as_system_duration();
 
-        loop {
-            let mut buf = [0_u8; 48];
+        poll_wait
+            .as_mut()
+            .reset(self.last_poll_sent + poll_interval);
+    }
 
-            tokio::select! {
-                () = &mut poll_wait => {
-                    let system_snapshot = *system_snapshots.borrow_and_update();
+    async fn handle_poll(&mut self, poll_wait: &mut Pin<&mut Sleep>) {
+        let system_snapshot = *self.channels.system_snapshots.read().await;
+        let packet = self.peer.generate_poll_message(system_snapshot);
 
+        // Sent a poll, so update waiting to match deadline of next
+        self.last_poll_sent = Instant::now();
+        self.update_poll_wait(poll_wait, system_snapshot);
```

Comment:
This was probably a pre-existing bug, but we really should also check peer acceptability here and communicate it if the peer has become unnacceptable to use for synchronization. If you dont want to fix that in this PR then we should make an issue for this.

Reply:
why is that needed here? can't we wait till the next packet we receive? What do we do when the peer is not acceptable any more (my guess is we just wait till the next packet)